crates/provider/src/heart.rs

```diff
@@ -448,38 +463,119 @@ impl<S> Heartbeat<S> {
             self.reap_at.insert(Instant::now() + timeout, to_watch.config.tx_hash);
         }
+        // Transaction may be confirmed already, check the lookbehind history first.
+        // If so, insert it into the waiting list.
+        for (block_height, txs) in self.past_blocks.iter().rev() {
+            if txs.contains(&to_watch.config.tx_hash) {
+                let confirmations = to_watch.config.required_confirmations;
+                let confirmed_at = *block_height + confirmations - 1;
+                let current_height = self.past_blocks.back().map(|(h, _)| *h).unwrap();
+
+                if confirmed_at <= current_height {
+                    to_watch.notify();
+                } else {
+                    debug!(tx=%to_watch.config.tx_hash, %block_height, confirmations, "adding to waiting list");
+                    self.waiting_confs.entry(confirmed_at).or_default().push(to_watch);
+                }
+                return;
+            }
+        }
+
         self.unconfirmed.insert(to_watch.config.tx_hash, to_watch);
     }
 
+    fn add_to_waiting_list(&mut self, watcher: TxWatcher, block_height: u64) {
+        let confirmations = watcher.config.required_confirmations;
+        debug!(tx=%watcher.config.tx_hash, %block_height, confirmations, "adding to waiting list");
+        self.waiting_confs.entry(block_height + confirmations - 1).or_default().push(watcher);
+    }
+
     /// Handle a new block by checking if any of the transactions we're
     /// watching are in it, and if so, notifying the watcher. Also updates
     /// the latest block.
     fn handle_new_block(&mut self, block: Block, latest: &watch::Sender<Option<Block>>) {
         // Blocks without numbers are ignored, as they're not part of the chain.
         let Some(block_height) = &block.header.number else { return };
 
+        // Add the block the lookbehind.
+        // The value is chosen arbitrarily to not have a huge memory footprint but still
+        // catch most cases where user subscribes for an already mined transaction.
+        // Note that we expect provider to check whether transaction is already mined
+        // before subscribing, so here we only need to consider time before sending a notification
+        // and processing it.
+        const MAX_BLOCKS_TO_RETAIN: usize = 10;
+        if self.past_blocks.len() >= MAX_BLOCKS_TO_RETAIN {
+            self.past_blocks.pop_front();
+        }
+        if let Some((last_height, _)) = self.past_blocks.back().as_ref() {
+            // Check that the chain is continuous.
+            if *last_height + 1 != *block_height {
+                // Remove all the transactions that were reset by the reorg.
+                warn!(%block_height, last_height, "reorg detected");
+                self.reap_reorg(*block_height);
+                // Remove past blocks that are now invalid.
+                self.past_blocks.retain(|(h, _)| h < block_height);
```

Comment:
maybe block_height <= last_height is more accurate here?

Reply:
No, `block_height` is the first block of the "new" chain after reorg, so transactions that belong to this block should also be removed. We will push the block with `block_height` right below.